app/src/main/java/ch/epfl/sdp/appart/database/firestoreservicehelpers/FirestoreAdHelper.java

```diff
@@ -0,0 +1,278 @@
+        Log.d("URI", "size" + uriList.size());
+        for (int i = 0; i < uriList.size(); i++) {
+            String name = FirebaseLayout.PHOTO_NAME + i + ".jpeg"; // TODO modify to support other extensions
+            actualRefs.add(name);
+            imagesUploadResults.add(imageHelper.putImage(uriList.get(i), name, storagePath));
+        }
+        // check whether any of the uploads failed
+        checkPhotosUpload(imagesUploadResults, imagesResult, newAdRef, cardRef, storage.getReference(storagePath));
+
+        // build and send card / ad
+        checkCardUpload(cardResult, ad, newAdRef, cardRef, null, actualRefs.get(0));
+        checkAdUpload(adResult, ad, newAdRef, cardRef, null, actualRefs);
+
+        // check if everything succeeded
+        finalizeAdUpload(result, imagesResult, adResult, cardResult, newAdRef);
+        return result;
+    }
+
+    /* <--- getAd private methods --->*/
+
+    private void getAndCheckImagesIds(CompletableFuture<List<String>> result,
+                                      String adId) {
+        db.collection(adsPath).document(adId).collection(AdLayout.PICTURES_DIRECTORY)
+                .get().addOnCompleteListener(task -> {
+            if (!task.isSuccessful()) {
+                result.completeExceptionally(new DatabaseServiceException("Failed to get pictures ids."));
+            } else {
+                result.complete(task.getResult().getDocuments().stream().map(documentSnapshot ->
+                        (String) documentSnapshot.get("id")).collect(Collectors.toList()));
+            }
+        });
+    }
+
+    private void getAndCheckPartialAd(CompletableFuture<Ad.AdBuilder> result,
+                                      String adId) {
+        db.collection(adsPath).document(adId).get().addOnCompleteListener(task -> {
+            if (task.isSuccessful()) {
+                DocumentSnapshot documentSnapshot = task.getResult();
+
+                Ad.AdBuilder builder = new Ad.AdBuilder()
+                        .withTitle((String) documentSnapshot.get(AdLayout.TITLE))
+                        .withPrice((long) documentSnapshot.get(AdLayout.PRICE))
+                        .withPricePeriod(PricePeriod.ALL.get(
+                                Math.toIntExact((long) documentSnapshot.get(AdLayout.PRICE_PERIOD))))
+                        .withStreet((String) documentSnapshot.get(AdLayout.STREET))
+                        .withCity((String) documentSnapshot.get(AdLayout.CITY))
+                        .withAdvertiserId((String) documentSnapshot.get(AdLayout.ADVERTISER_ID))
+                        .withDescription((String) documentSnapshot.get(AdLayout.DESCRIPTION))
+                        .hasVRTour((boolean) documentSnapshot.get(AdLayout.VR_TOUR));
+
+                result.complete(builder);
+            } else {
+                result.completeExceptionally(new DatabaseServiceException(Objects.requireNonNull(
+                        task.getException()).getMessage()));
```

Comment:
Since we are sure the task failed, I don't think the Object.requireNonNull is needed. Bet we can keept it.

Reply:
The linter asks for it, so I would leave it